jenkins/deploy-clio-server.sh

```diff
@@ -1,210 +1,256 @@
+Steps:
+  1. Render clio-server ctmpls into configs for the given environment
+  2. Copy rendered configs to the host
+  3. Stop any currently-running clio container network on the host
+  4. Start a new network using the new configs and docker image
+  5. Verify that the new Clio server reports OK health and the expected version
+  6. Roll back the deploy if the new server fails to report its health or version
+
+Note: requires the following ENV vars to be set to non-null values:
+
+  ENV:            Environment to deploy to, one of "dev", "staging", or "prod"
+  CLIO_HOST_NAME: Name of the host within the network specified by 'ENV' to deploy to,
+                  for example: "clio101", "clio201", "clio3", "clio"
+
+Also requires that a file containing a vault token be placed at one of the following paths:
+
+  /etc/vault-token-dsde
+  ${HOME}/.vault-token
+EOF
+}
+
+check_env_args() {
+  local ret=0
+
+  if [ -z "$ENV" ]; then
+    >&2 echo "Error: ENV not set!"
+    ret=1
+  else
+    case "$ENV" in
+      "dev"|"staging"|"prod")
+        ;;
+      *)
+        >&2 echo "Error: Invalid environment '$ENV'!"
+        ret=1
+        ;;
+    esac
+  fi
+
+  if [ -z "$CLIO_HOST_NAME" ]; then
+    >&2 echo "Error: CLIO_HOST_NAME not set!"
+    ret=1
+  fi
+
+  if [ $ret -ne 0 ]; then
+    usage
+    exit 1
+  fi
+}
+
+render_ctmpls() {
+  local -r docker_tag=$1 clio_domain_name=$2 tmp=$3 app_dir=$4 host_log_dir=$5 pos_log_dir=$6
+
+  local -r ctmpl_dir="${CLIO_DIR}/clio-server/config"
+  local vault_token_file
+
+  if [ ! -d "$ctmpl_dir" ]; then
+    >&2 echo "Error: Missing ctmpl source directory ($ctmpl_dir) - Was it accidentally deleted?"
+    exit 1
+  fi
+
+  # Try to find the vault token location.
+  if [ -f /etc/vault-token-dsde ]; then
+    vault_token_file=/etc/vault-token-dsde
+  elif [ -f ${HOME}/.vault-token ]; then
+    vault_token_file=${HOME}/.vault-token
+  else
+    >&2 echo "Error: Vault token file not found at /etc/vault-token-dsde or ${HOME}/.vault-token"
+    exit 1
+  fi
+
+  # Location in the containers where configs will be mounted.
+  local -r clio_conf_dir=/etc
+  # Location in the Clio container where logs will be mounted.
+  local -r clio_log_dir=/logs
+
+  # Name of the application config used by the Clio instance.
+  local -r clio_app_conf=clio.conf
+  # Name of the logback config used by the Clio instance.
+  local -r clio_logback_conf=clio-logback.xml
+  # Name of the service account JSON file used by the Clio instance.
+  local -r clio_service_account_json=clio-service-account.json
+
+  # Port to expose for Clio in the container.
+  local -r container_clio_port=8080
+
+  # Populate an env file with variables for rendering configs.
+  local -r ctmpl_env_file="${tmp}/env-vars.txt"
+  for var in ENV clio_domain_name docker_tag app_dir clio_conf_dir host_log_dir clio_log_dir pos_log_dir clio_app_conf clio_logback_conf clio_service_account_json container_clio_port; do
+    echo "${var^^}=${!var}" >> ${ctmpl_env_file}
```

Comment:
Comment here to explain bash magic

Reply:
I agree! The original code `echo "${var}=${!var}"` made sense to me, not sure what `^^` is doing here.